MapleServer2/PacketHandlers/Game/InteractObjectHandler.cs

```diff
@@ -61,9 +83,43 @@ private static void HandleUse(GameSession session, PacketReader packet)
                     session.Send(QuestPacket.CompleteQuest(item.Basic.QuestID));
                     break;
                 }
+                session.Send(InteractActorPacket.UseObject(actor));
+            }
+            if (actor.Type == InteractActorType.Gathering)
+            {
+                RecipeMetadata recipe = RecipeMetadataStorage.GetRecipe(actor.RecipeId);
+                long requireMastery = int.Parse(recipe.RequireMastery);
+                Enums.MasteryType type = (Enums.MasteryType) int.Parse(recipe.MasteryType);
+                long currentMastery = 0;
+                int stamina = 10;
+                MasteryExp mastery = session.Player.Levels.MasteryExp.FirstOrDefault(x => x.Type == (byte) type);
+                if (mastery != null)
+                {
+                    currentMastery = mastery.CurrentExp;
+                    stamina = mastery.Stamina;
+                }
+                List<RecipeItem> items = RecipeMetadataStorage.GetResult(recipe);
+                Random rand = new Random();
+                InteractObjecResult result = InteractObjecResult.Fail;
+                foreach (RecipeItem item in items)
+                {
+                    if ((currentMastery >= requireMastery)
+                        && (rand.Next(100) < (RarityChance[item.Rarity] * stamina / 10)))
+                    {
+                        session.FieldManager.AddItem(session, new Item(item.Id));
+                        result = InteractObjecResult.Success;
+                    }
+                }
+                if (result == InteractObjecResult.Success)
+                {
+                    session.Player.Levels.GainMasteryExp(type, recipe.RewardMastery);
+                }
+                if ((mastery != null) && (mastery.Stamina > 0))
+                {
+                    mastery.Stamina -= 1;
+                }
+                session.Send(InteractActorPacket.UseObject(actor, (short) result));
```

Comment:
I think having `UseObject()` on line 121 and 86 is not needed could be just one line to both cases, it also breaks interacting with extractor.